source/vstd/pcm/frac.rs

```diff
@@ -0,0 +1,272 @@
+use super::super::modes::*;
+use super::super::prelude::*;
+use super::*;
+
+// This implements a resource for fractional ownership of a ghost variable.
+// The fractions are represented as some number out of a compile-time const
+// TOTAL value; you can have any fractions from 1 up to TOTAL, and if you
+// have TOTAL, you can update the ghost variable.
+
+verus! {
+
+// Too bad that `nat` and `int` are forbidden as the type of a const generic parameter
+enum FractionalCarrier<T, const TOTAL: u64> {
+    Value { v: T, n: int },
+    Empty,
+    Invalid,
+}
+
+impl<T, const TOTAL: u64> FractionalCarrier<T, TOTAL> {
+    spec fn new(v: T) -> Self {
+        FractionalCarrier::Value { v: v, n: TOTAL as int }
+    }
+}
+
+impl<T, const TOTAL: u64> PCM for FractionalCarrier<T, TOTAL> {
+    closed spec fn valid(self) -> bool {
+        match self {
+            FractionalCarrier::Invalid => false,
+            FractionalCarrier::Empty => true,
+            FractionalCarrier::Value { v: v, n: n } => 0 < n <= TOTAL,
+        }
+    }
+
+    closed spec fn op(self, other: Self) -> Self {
+        match self {
+            FractionalCarrier::Invalid => FractionalCarrier::Invalid,
+            FractionalCarrier::Empty => other,
+            FractionalCarrier::Value { v: sv, n: sn } => match other {
+                FractionalCarrier::Invalid => FractionalCarrier::Invalid,
+                FractionalCarrier::Empty => self,
+                FractionalCarrier::Value { v: ov, n: on } => {
+                    if sv != ov {
+                        FractionalCarrier::Invalid
+                    } else if sn <= 0 || on <= 0 {
+                        FractionalCarrier::Invalid
+                    } else {
+                        FractionalCarrier::Value { v: sv, n: sn + on }
+                    }
+                },
+            },
+        }
+    }
+
+    closed spec fn unit() -> Self {
+        FractionalCarrier::Empty
+    }
+
+    proof fn closed_under_incl(a: Self, b: Self) {
+    }
+
+    proof fn commutative(a: Self, b: Self) {
+    }
+
+    proof fn associative(a: Self, b: Self, c: Self) {
+    }
+
+    proof fn op_unit(a: Self) {
+    }
+
+    proof fn unit_valid() {
+    }
+}
+
+pub struct Frac<T, const TOTAL: u64 = 2> {
+    r: Resource<FractionalCarrier<T, TOTAL>>,
+}
+
+impl<T, const TOTAL: u64> Frac<T, TOTAL> {
+    pub closed spec fn inv(self) -> bool {
```

Comment:
you can probably use the `type_invariant` attribute here, so your library doesn't need to include self.inv() in the requires and ensures clauses.

Reply:
I tried doing this originally, but ran into this problem:

```
% cat example.rs
use vstd::prelude::*;

verus! {
    struct S<const N: u64> {
    }

    impl<const N: u64> S<N> {
        #[verifier::type_invariant]
        pub closed spec fn inv(self) -> bool { true }
    }
}
% verus --crate-type=lib example.rs
error: external_type_specification: Const params not yet supported
 --> example.rs:9:9
  |
9 |         pub closed spec fn inv(self) -> bool { true }
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 1 previous error
```

Reply:
this should be fixed with 0f22710fc04740132d0466ce9dd223011abf150a

Reply:
Thanks!  Do you have any suggestions for how to use the type invariant of a mutable reference?  Do we need to add a version of `use_type_invariant()` that works on a `tracked &mut T`?

```
% cat example.rs
use vstd::prelude::*;

verus! {
    struct S {}

    impl S {
        #[verifier::type_invariant]
        pub closed spec fn inv(self) -> bool { true }

        proof fn x(tracked &mut self) {
            use_type_invariant(self);
        }
    }
}
% verus --crate-type=lib example.rs
error: The verifier does not yet support the following Rust feature: &mut types, except in special cases
  --> example.rs:11:32
   |
11 |             use_type_invariant(self);
   |                                ^^^^

error: aborting due to 1 previous error
```

Reply:
Ah, turns out `use_type_invariant(&*self)` works in this case.  Updated the PR to use type invariants and avoid requiring an explicit `frac.inv()`.  Thanks for the suggestion!